docs/keps/20200619-kubefed-pull-reconciliation.md

```diff
@@ -0,0 +1,362 @@
+all the kubefed clusters.
+
+The future of `ServiceDNSRecord` controller is something to decide with the community.
+It goes one step beyond by dealing with networking considerations when talking to apps deployed across clusters.
+
+On the other hand, the controllers in charge of the scheduler preferences remains as part of the control-plane.
+The enforcement of these preferences rely on the control-plane and daemons, so nothing would change there.
+
+The rest of kubefed cluster related operations would be managed by controllers in the control-plane.
+
+### Controller Resource Propagation Status
+
+The creation of a federated resource triggers an asynchronous process that updates
+the state of the operation once applied to all the Kubefed clusters.
+
+These actions trigger operations in the control-plane and kubefed clusters.
+
+The following list represents the different steps when creating a resource:
+
+1. Create federated resource.
+2. Update status of the federated resource to reflect the initiation of the propagation (e.g. `Propagating`).
+3. A propagation reconcile loop is triggered to update the the status of this resource until the completion of this operation.
+4. As part of the reconciliation loop, the system checks which kubefed clusters
+are specified as part of the `placement` for this resource. Next, the system verifies
+if the federated resource exists, if it does the new status is changed to `Propagated`, and the
+reconciliation loop ends successfully. Otherwise the reconcile request is re-queue
+until the resource is created in the target cluster or an error is reported (`CreationTimedOut` or so).
+
+A similar process applies to deletion and update operations on the federated resources.
+However, in a deletion of a federated resource, this operation is considered completed when the resource
+does not exist on any of the allocated clusters.
+
+This approach assumes the control-plane polls the status of the `notready` federated
+resources from the kubefed clusters using an endpoint exposed by the daemons.
+However there is another alternative that was not presented yet in this document.
+The daemons could alternatively have `write` permissions to save the status of their federated resources to
+keep the state up-to-date for each resource.
+
+### Kubefed Security Concerns
+
+With this new approach, the kubefed daemon only exposes read permissions to ensure
+the reconciliation of the current status of the federated resources.
+Likewise the daemons require a read access to the federated types in the control-plane cluster to list the available
+federated types and watch the resources that have to be federated on each cluster.
+This differs from the current architecture where the control-plane has write access
+to the kubefed clusters.
+
+Consequently the `control-plane` and `kubefed daemons` need to communicate between themselves.
+A bi-directional communication is required, consequently a secure trust communication should
```

Comment:
As per my comment about getting the cluster status and resource status details from the cluster API endpoint the bi directional access might not be needed.

Reply:
Yeah, we could define an iterative approach for now. But I'd like you to understand the intention behind the kubefed daemons exposing the endpoints as done for the vkubelets.

Reply:
@irfanurrehman the endpoints approach has been replaced by a `fully` decentralized approach